refactor/jobs-screen

mohammed-akkad opened the pull request:
This pull request introduces robust error handling for network connectivity issues and improves the user experience in the craftsman jobs feature. It adds explicit checks for internet connection, provides user feedback through error messages and snackbars, and refactors the ViewModel for cleaner separation of concerns and better maintainability.

**Network connectivity handling:**

* Added `NetworkConnectionChecker` to `RequestsRepositoryImpl` and now throw `NoInternetConnectionException` if there is no internet when fetching craftsman requests, ensuring the app responds gracefully to offline scenarios. 

* Updated all relevant ViewModel operations to handle `NoInternetConnectionException`, updating UI state to show appropriate messages or snackbars when the device is offline.

**UI/UX improvements:**

* Added new UI state fields and error messages for clearer feedback to the user, including handling of error snackbars and retry/dismiss actions. 

**Code structure and maintainability:**

* Refactored ViewModel methods to extract error and success handlers, improving readability and separation of concerns. 

**UI enhancements:**

* Began integrating `AnimatedVisibility` for improved UI feedback (preparation for future UI changes). (`MyJobsScreen.kt`)

mohammed-akkad committed:
update strings

mohammed-akkad committed:
refactor jobs screen

mohammed-akkad committed:
Merge branch 'develop' into refactor/jobs-screen

Joseph-Sameh-0 commented on app/src/main/java/com/paris_2/san3a/presentation/shared/components/PlaceHolderScreen.kt:
```diff
@@ -52,7 +53,7 @@ fun PlaceHolderScreen(
                 text = stringResource(description),
                 style = Theme.textStyle.body.medium.regular,
-                modifier = Modifier.padding(bottom = 24.dp),
+                modifier = Modifier.width(280.dp).padding(bottom = 24.dp),
```
why this fixed width?

mohammed-akkad replied:
it is like design, and without it, text take fill max width from start to end

Joseph-Sameh-0 replied:
you can make horizontal padding instead of this but make it take additional width if available for large devices

Joseph-Sameh-0 replied:
I Suggest to don't change the component 
make the screen control it

Joseph-Sameh-0 committed:
Refactor: Update logo path data and logging tag in MyJobsCraftsmanViewModel

ahmedNaser7 committed:
feat: update app launcher icon for dark and light theme

This commit updates the app's launcher icon by:
- Replacing the previous foreground and background drawables with new ones.
- Introducing `ic_san3a_launcher.xml` for the foreground vector graphic.
- Adding `ic_san3a_launcher_inset.xml` to apply an inset to the foreground.
- Defining `ic_launcher_background` color resource for both light and dark themes.
- Updating `AndroidManifest.xml` to use `ic_launcher` for both `android:icon` and `android:roundIcon`.

ahmedNaser7 committed:
Merge branch 'develop' into refactor/jobs-screen

# Conflicts:
#	app/src/main/AndroidManifest.xml

Joseph-Sameh-0 committed:
Fix: Handle empty lists and improve messaging UI

- Add `GetCraftsmanRequestsException` for handling errors when fetching craftsman requests.
- Handle cases where `requestIds`, `docsList`, or `relatedJobs` are empty in data sources to prevent crashes.
- Ensure `isSeen` is not null before using it in `MessageMeta`.
- Modify `MessageTextField` to have a max height and fix background shape.
- Update `MessagesScreen` and `ChatList` to use default values for `theOtherUserName` and `theOtherProfilePhoto` if they are null.
- Handle null or empty `theOtherProfilePhoto` in `MessagesState`.
- Improve error handling in `MessagesViewModel` when fetching user chat info.
- Improve error handling and loading states in `MessagesDetailsViewModel`.
- Show sending message UI in `MessageDetails` screen.
- Scroll to the bottom of the message list when a new message is sent or received.
- Add profile image to the `MessageDetails` screen top bar.

Joseph-Sameh-0 committed:
Merge branch 'refactor/jobs-screen' of https://github.com/Paris-Squad-S2/San3a into refactor/jobs-screen

Joseph-Sameh-0 committed:
Refactor: Handle null and empty lists in ViewModels and update MyJobsScreen padding

- Update various ViewModels to handle null or empty lists from use cases by providing default empty collections (lists/maps) or default values. This prevents potential null pointer exceptions and ensures UI stability.
  - `CustomerHomeViewModel`: `mostRequestedServices`
  - `MoreViewModel`: `userRating`
  - `MyRequestCustomerViewModel`: `ongoing`, `completed`, `canceled` requests
  - `NotificationViewModel`: `notifications`
  - `CustomerRequestDetailsRequestDetailsViewModel`: `offers`
  - `MyServiceViewModel`: `services`
  - `MessagesDetailsViewModel`: `messages`
  - `AccountViewModel`: `services`
  - `CraftsmanRequestDetailsViewModel`: `offers`
  - `MyJobsCraftsmanViewModel`: `result` (for craftsman offers)
  - `CraftsmanHomeViewModel`: `userServices`, `stats`, `recentRelatedJobs`, `offersCount`, `availableJobs`
  - `MessagesViewModel`: `chatsMap`
- Modify `BaseViewModel`'s `tryToObserve` function to call `onEach` with `null` if the observed Flow does not emit any items.
- Adjust padding in `MyJobsScreen` for the title and action icon.

Joseph-Sameh-0 committed:
Refactor: Handle null and empty lists in ViewModels and update MyJobsScreen padding

- Update various ViewModels to handle null or empty lists from use cases by providing default empty collections (lists/maps) or default values. This prevents potential null pointer exceptions and ensures UI stability.
  - `CustomerHomeViewModel`: `mostRequestedServices`
  - `MoreViewModel`: `userRating`
  - `MyRequestCustomerViewModel`: `ongoing`, `completed`, `canceled` requests
  - `NotificationViewModel`: `notifications`
  - `CustomerRequestDetailsRequestDetailsViewModel`: `offers`
  - `MyServiceViewModel`: `services`
  - `MessagesDetailsViewModel`: `messages`
  - `AccountViewModel`: `services`
  - `CraftsmanRequestDetailsViewModel`: `offers`
  - `MyJobsCraftsmanViewModel`: `result` (for craftsman offers)
  - `CraftsmanHomeViewModel`: `userServices`, `stats`, `recentRelatedJobs`, `offersCount`, `availableJobs`
  - `MessagesViewModel`: `chatsMap`
- Modify `BaseViewModel`'s `tryToObserve` function to call `onEach` with `null` if the observed Flow does not emit any items.
- Adjust padding in `MyJobsScreen` for the title and action icon.

Joseph-Sameh-0 committed:
Refactor: Remove unused import in RequestsRepositoryImpl
